Gems/AzQtComponentsForPython/CMakeLists.txt

```diff
@@ -0,0 +1,224 @@
+
+# Check we have a valid version of Python.
+if(NOT python_interpreter)
+    find_program(python_interpreter "python")
+endif()
+
+if(NOT python_interpreter)
+    message(FATAL_ERROR "Error: Failed to locate Python installation.")
+endif()
+
+set(LIB_EXTENSION "cp310-win_amd64")
+
+# Find the paths for the pyside2/shiboken modules.
+ly_get_package_association("pyside2" PYSIDE_PACKAGE_NAME)
+ly_package_get_target_folder(${PYSIDE_PACKAGE_NAME} PYSIDE2_DOWNLOAD_LOCATION)
+set(PYSIDE2_MODULE_DIR ${PYSIDE2_DOWNLOAD_LOCATION}/${PYSIDE_PACKAGE_NAME})
+
+# Find the qt package location.
+ly_get_package_association("Qt" QT_PACKAGE_NAME)
+ly_package_get_target_folder(${QT_PACKAGE_NAME} QT_DOWNLOAD_LOCATION)
+set(QT_MODULE_DIR ${QT_DOWNLOAD_LOCATION}/${QT_PACKAGE_NAME}/qt)
+
+set(PYSIDE2_PATH "${PYSIDE2_MODULE_DIR}/pyside2")
+set(INCLUDE_DIR "${PYSIDE2_PATH}/include")
+set(PYSIDE2_SHARE_PATH "${PYSIDE2_PATH}/share/PySide2")
+set(PYSIDE2_TYPESYSTEM_PATH "${PYSIDE2_SHARE_PATH}/typesystems")
+set(PYSIDE2_BIN_PATH "${PYSIDE2_PATH}/bin")
+set(PYSIDE2_INCLUDE_DIR ${INCLUDE_DIR}/PySide2)
+set(SHIBOKEN_GENERATOR_EXE_PATH "${PYSIDE2_BIN_PATH}/shiboken2${CMAKE_EXECUTABLE_SUFFIX}") 
+
+if(NOT EXISTS ${SHIBOKEN_GENERATOR_EXE_PATH})
+    message(FATAL_ERROR "Shiboken executable not found at path: ${SHIBOKEN_GENERATOR_EXE_PATH}")
+endif()
+  
+set(SHIBOKEN_MODULE_PATH "${PYSIDE2_MODULE_DIR}/pyside2/lib/site-packages/shiboken2")
+
+set(SHIBOKEN_INCLUDE_DIR "${INCLUDE_DIR}/shiboken2")
+set(SHIBOKEN_SHARED_LIBRARIES "${SHIBOKEN_MODULE_PATH}/shiboken2.abi3.lib")
+set(SHIBOKEN_DLL "${SHIBOKEN_MODULE_PATH}/shiboken2.abi3.dll")
+set(SHIBOKEN_SHARED_LIBRARIES_D "${SHIBOKEN_MODULE_PATH}/shiboken2_d.${LIB_EXTENSION}.lib")
+set(SHIBOKEN_DLL_D "${SHIBOKEN_MODULE_PATH}/shiboken2_d.${LIB_EXTENSION}.dll")
+
+set(PYSIDE2_LIB_DIR ${PYSIDE2_PATH}/lib/site-packages/PySide2)
+
+set(PYSIDE2_SHARED_LIBRARIES "${PYSIDE2_LIB_DIR}/pyside2.abi3.lib")
+set(PYSIDE2_SHARED_LIBRARIES_D "${PYSIDE2_LIB_DIR}/pyside2_d.${LIB_EXTENSION}.lib")
+
+set(PYSIDE2_DLL "${PYSIDE2_LIB_DIR}/pyside2.abi3.dll")
+set(PYSIDE2_DLL_D "${PYSIDE2_LIB_DIR}/pyside2_d.${LIB_EXTENSION}.dll")
+  
+set(CMAKE_SKIP_BUILD_RPATH FALSE)
+set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
+set(CMAKE_INSTALL_RPATH ${SHIBOKEN_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR})
+set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+
+# Define the shared library for which we will create bindings.
+set_property(TARGET ${AZQTCOMPONENTS_LIBRARY} PROPERTY PREFIX "")
+
+# Ensure the generated wrapper files have the correct header.
+set(LICENSE_FILE ${CMAKE_CURRENT_LIST_DIR}/license_header.txt)
+
+# Set up the options to pass to shiboken.
+set(shiboken_options --generator-set=shiboken --enable-parent-ctor-heuristic
+    --enable-pyside-extensions --enable-return-value-heuristic --use-isnull-as-nb_nonzero
+    --avoid-protected-hack --language-level=c++17
+    --license-file=${LICENSE_FILE}
+    ${INCLUDES}
+    -I${CMAKE_SOURCE_DIR}
+    -I${CMAKE_CURRENT_LIST_DIR}/..
```

Comment:
Were these include directories randomly added.
This should add the include directories from the AzQtComponents TARGET and it's dependencies through the [INCLUDE_DIRECTORIES](https://cmake.org/cmake/help/latest/prop_tgt/INCLUDE_DIRECTORIES.html?highlight=include_directories) TARGET property.
Afterwards It should also only need to add the directory with containing the `bindings_azqtcomponents.h` file to the include paths.

Reply:
Agreed, tidying up needed. I'll look into using INCLUDE_DIRECTORIES, but I'm not sure they'll get correctly passed through the the shiboken exe that way.

Reply:
> It feels like theres a giant piece missing from this (or perhaps its already there and I'm misunderstanding?)
> 
> I expect to see something like a shiboken 3p package containing the shiboken executables. I expect to see a 3p package for each official platform (Mac, windows, linux).
> 
> I expect the gem to depend on the 3p package to get targets like "shiboken" executable itself, we likely wouldn't check these executables into this gem at all. Is there maybe one, and I'm just missing it here and thus lack context? If there is one, it should expose pretty much everything we need so that not just O3DE, but anyone else who's writing a gem can issue a custom step using it.
> 
> Then the gem-specific cmakelists.txt for AzQtComponentsForPython would not contain any platform or shiboken specific stuff... just invoking a command added by the shiboken 3p... right?
> 
> In other words to support an external 3p like shiboken I'd expect us to make a real 3p package for it (or use externalproject or fetchcontent) to make a set of targets that represents the stuff inside it as well as any custom commands or exported directories.
> 
> Then I'd expect to see this particular gem to just "Use" that integration of that 3p to list its inputs to the generator and outputs and so on, not contain tons of shiboken specific configuration itself. Then any other gems which also need to expose stuff via shiboken would also "use" that 3p integation, and invoke the same commands on their own files that need generator treatement.

Shiboken is now included in the PySide 3p package which helpes, but there's still quite a bit of setup involved: maybe the FindPySide2 script should be modified to handle setting up some of it.